### PR TITLE
Update pool url on redirect

### PR DESCRIPTION
**Pull request by felixbrucker**
### Purpose:
Support the migration of pool farmer api urls: when a pool changes its farmer api url, the client persists the new url to be able to continue to communicate with the pool should the old url stop working in the future.

### Current Behavior:
Pools can already setup redirects to a new farmer api url and they will get used correctly, however the new url is not persisted by the client.

### New Behavior:
When a url change is detected the new url is persisted.

### CODE REVIEW

**Comment by emlowe** [comment]
Needs a test to execise this code - take a look at `test_farmer_pool_response`

**Comment by felixbrucker** [comment]
Unfortunately i know too little of how the mocking in pytest works for the following scenario:
- mock GET for one url with a 307 redirect and location set to another url
- mock a second GET for the other url with a json response

It seems the currently used patch just overwrites what the specific `aiohttp.ClientSession.*` method returns, i'd need something like https://github.com/pnuckowski/aioresponses probably, or is something like this achievable with some kind of callback style return value where i do the matching myself?

**Comment by felixbrucker** [comment]
@emlowe added tests 👍

File: chia/farmer/farmer.py [hunk 1]
```diff
@@ -327,16 +334,19 @@ async def plot_sync_callback(self, peer_id: bytes32, delta: Optional[Delta]) ->
...
-                    f"{pool_config.pool_url}/pool_info", ssl=ssl_context_for_root(get_mozilla_ca_crt(), log=self.log)
-                ) as resp:
+                url = f"{pool_config.pool_url}/pool_info"
+                async with session.get(url, ssl=ssl_context_for_root(get_mozilla_ca_crt(), log=self.log)) as resp:
                     if resp.ok:
                         response: Dict[str, Any] = json.loads(await resp.text())
                         self.log.info(f"GET /pool_info response: {response}")
-                        return response
+                        new_pool_url: Optional[str] = None
+                        if resp.url != url:
```
Comment (arvidn) [inline chia/farmer/farmer.py:346]: HTTP redirects come in different flavours. Only in the cases of `301 Moved Permanently` and `308 Permanent Redirect` does it seem reasonable to persist the new URL, and use it for future queries/posts.
e.g. `302 Found` (described as "moved temporarily") would not make sense to persist.

Comment (felixbrucker) [reply chia/farmer/farmer.py:346]: Sure, i can restrict by status code

File: tests/farmer_harvester/test_farmer.py [hunk 2]
```diff
@@ -908,3 +911,287 @@ def assert_pool_errors_24h() -> None:
...
+        "pool_url": "localhost",
+        "launcher_id": "ae4ef3b9bfe68949691281a015a9c16630fc8f66d48c19ca548fb80768791afa",
+        "target_puzzle_hash": "344587cf06a39db471d2cc027504e8688a0a67cce961253500c956c73603fd58",
+    }
+    for key, value in overrides.items():
+        pool_list_entry[key] = value
+    return pool_list_entry
+
+
+def make_pool_info(overrides: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (emlowe) [inline tests/farmer_harvester/test_farmer.py:930]: pretty minor but this function allows for overrides to be passed in, but it's only called with an empty dict I think. So you can simplify by removing the overrides

(it was flagged in coverage as not being used)

Comment (felixbrucker) [reply tests/farmer_harvester/test_farmer.py]: That makes sense, done 👍

**Comment by emlowe** [comment]
close and reopen to pickup hopefully some CI fixes

**Review by emlowe — APPROVED** [review]

**Comment by emlowe** [comment]
coverage diff exemption

**Pull request merged**